jumanji/environments/routing/pacman/utils.py

```diff
@@ -0,0 +1,470 @@
+        move_right = lambda position: (position[1], position[0] + 1)
+        move_down = lambda position: (position[1] + 1, position[0])
+        no_op = lambda position: (position[1], position[0])
+
+        new_pos_row, new_pos_col = jax.lax.switch(
+            chosen_action,
+            [move_left, move_up, move_right, move_down, no_op],
+            position,
+        )
+
+        # If using the teleporter shift to the other side of the map
+        path = jnp.array([new_pos_col % state.y_size, new_pos_row % state.x_size])
+        path = jax.lax.cond(ghost_start <= 0, lambda: path, lambda: position)
+
+        return path, chosen_action
+
+    ghost_nums = jnp.array([0, 1, 2, 3])
+    ghost_init_targets = state.ghost_init_targets
+    ghost_paths, ghost_acts = jax.vmap(move, in_axes=(0, 0, 0, 0, None, 0, 0, 0, 0))(
+        jnp.array(ghost_pos),
+        ghost_keys,
+        actions,
+        ghost_nums,
+        player_pos,
+        jnp.array(ghost_init_targets),
+        jnp.array(state.old_ghost_locations),
+        start_time,
+        scatter_targets,
+    )
+
+    return ghost_paths, ghost_acts, key
+
+
+def check_ghost_wall_collisions(
+    state: State,
+    new_player_pos: chex.Array,
+    ghost_num: int,
+    pacman_pos: Position,
+    init_target: chex.Array,
+    old_ghost_locations: chex.Array,
+    scatter_target: chex.Array,
+) -> Tuple[bool, chex.Array, chex.Array]:
+
+    """
+    Determine valid move for the ghost.
+
+    Args:
+        state: 'State` object corresponding to the new state of the environment.
+        new_player_pos: the position of the player after the last action.
+        ghost_num: the id of the ghost being moved.
+        pacman_pos: the current position of the player.
+        init_target: number of steps left of initial movement.
+        old_ghost_locations: the position of the ghost in the previous step.
+        scatter_target: the scatter target of the ghost being moved.
+
+    Returns:
+        collision: a boolean indicating if the player has moved into a wall.
+        valid_locs: jax array of all possible valid locations.
+        masked_dist: jax array of valid target locations.
+    """
+
+    grid = jnp.array(state.grid)
+    pacman_pos = Position(x=pacman_pos.y, y=pacman_pos.x)
+    pac_dir = state.last_direction
+    is_scared = state.frightened_state_time  # [0]
+    ghost_init_steps = state.ghost_init_steps
+
+    # Get all possible positions
+    old_ghost_location = jnp.array([old_ghost_locations[1], old_ghost_locations[0]])
+    a0 = new_player_pos[1], new_player_pos[0] - 1
+    a1 = new_player_pos[1] - 1, new_player_pos[0]
+    a2 = new_player_pos[1], new_player_pos[0] + 1
+    a3 = new_player_pos[1] + 1, new_player_pos[0]
+    ghost_p = jnp.array([a0, a1, a2, a3])
+
+    # Block old paths so ghosts don't backtrack
+    ghost_mask = jnp.any(ghost_p != old_ghost_location, axis=1)
+
+    # For ghost 0: Move to closest tile to pacman
```

Comment:
Very cool that you got all the same behaviors as the original game!

Reply:
Thanks! @DriesSmit stopped me from making this convoluted